File: 467-unique-substrings-in-wraparound-string/ken/467.py
class Solution(object):
    def findSubstringInWraproundString(self, p):
        """
        :type p: str
        :rtype: int
        """

        def findSubStrLenAtIdx(s, idx):
            prev_c = s[idx]
            cnt = 1

            for i in xrange(idx + 1, len(s)):
                curr_c = s[i]

                if (ord(curr_c) - ord(prev_c) != 1) and \
                   not (prev_c == 'z' and curr_c == 'a'):
                    break

                prev_c = curr_c
                cnt += 1

            return cnt

        maxStrLenList = [0 for i in xrange(ord('a'), ord('z') + 1)]
        i = 0
        while i < len(p):
            substrLen = findSubStrLenAtIdx(p, i)

            for j in xrange(substrLen):
                idx = ord(p[i + j]) - ord('a')
                newLen = substrLen - j
                maxStrLenList[idx] = max(newLen, maxStrLenList[idx])

            i += substrLen

        return sum(maxStrLenList)

File: 467-unique-substrings-in-wraparound-string/leochang/467.cpp
class Solution {
public:
	int findSubstringInWraproundString(string p) {
		if (p.size() == 0) return 0;

		int sum = 0, max_len = 1;
		vector<int> dp(26, 0);

		dp[p[0] - 'a'] = max_len;

		for (int i = 1; i < p.size(); i++) {
			int tmp = p[i] - 'a';
			if (((p[i-1] - 'a')  + 1) % 26 == tmp) {
				max_len++;
			} else {
				max_len = 1;
			}
			dp[tmp] = max(max_len, dp[tmp]);
		}

		for (int i = 0; i < 26; i++) {
			sum += dp[i];
		}

		return sum;
	}
};
